File: src/coexpnet.cpp
// Created by Sajal Kumar
// Copyright (c) NMSU Song lab

#include "Silhouette.h"
#include "methods.h"
#include <boost/thread.hpp>
#include <armadillo>

/*
 * Worker -- Parallel discrete coexpression network builder that works with continuous values.
 * 
 * Parameters
 * 
 * children: A continuous sample x gene matrix representing 
 * discrete levels of gene expression at given samples. This matrix
 * contains all child genes.
 * 
 * parents: A continuous sample x gene matrix representing 
 * discrete levels of gene expression at given samples. This matrix
 * contains all parent genes.
 * 
 * c_names: a vector of strings containing gene names for the children matrix.
 * 
 * c_indices: a vector of integers representing the memory location in 'sc' that the 
 * worker process should work on.
 * 
 * p_names: a vector of strings containing gene names for the parents matrix.
 * 
 * k: a vector of integers that can contain the following:
 * i) k with size two: the minimum and maximum number of discrete levels allowed.
 * ii) k with size one: the maximum number of discrete levels allowed (minimum would be assumed to be 2).
 * iii) k with size n (where n is nparents x nchildren): the number of discrete levels to be used for each
 * interaction.
 *
 * sc: An instance of stat_collector to gather results.
 * 
 */

void coexpnet_thread(const frame<double> & children,
                     const frame<double> & parents,
                     const vec<std::string> & c_names,
                     vec<int> c_indices,
                     const vec<std::string> & p_names,
                     const vec<int> & k,
                     vec<stat_collector> & sc){

    // dims for input parents
    int nparents = parents.size();
    int nchildren_t = c_indices.size();

    // temporary contingency table
    frame<int> temp_tab;

    // temporary frame to store gene pairs
    frame<double> temp_RData;

    // temporary frame to store discretized data
    frame<int> temp_Ddata;

    // temporary arma mat for all pairs
    arma::mat temp_pair_data(2, parents[0].size());

    // temporary vector to store cluster assignments
    vec<int> temp_classn;

    // temporary frame to store distance matrix
    frame<double> dist_mat;

    // temporary variables to store silhouette scores (and best k)
    frame<double> sil_scores;
    double mean_sil_score;
    int best_k;
    double best_sil_score;

    // temporary variables for stat collection
    double stat, estimate;
    ldouble p_value = 1;
    int effind, temp_k_min, temp_k_max;

    for(int c=0; c<nchildren_t; c++) {

        // collect result for all parents except which c_names[c_indices[c]] == p_names[i]
        for(int i=0; i<nparents; i++){

            // effective index
            effind = (nparents * c_indices[c]) + i;

            if(c_names[c_indices[c]] == p_names[i]){

                // Not applicable -- store worst result
                fill_stat_collector(sc[effind], i, c_indices[c], p_names[i], c_names[c_indices[c]], 1, 0);

            } else {

                // create data
                create_pair_data(temp_RData, parents[i], children[c_indices[c]]);

                for(int mi=0; mi<temp_RData[0].size(); mi++){
                    for(int mj=0; mj<temp_RData.size(); mj++){
                        temp_pair_data(mi,mj) = temp_RData[mj][mi];
                    }
                }

                // if size of k is 2 then it specifies the maximum and minimum range.
                // if size of k is 1 then it specifies the maximum range
                if(k.size() == 2 || k.size() == 1){

                    // get distance matrix
                    dist_mat = DistMat(temp_RData);

                    // default k
                    temp_k_min = 2;
                    temp_k_max = 2;

                    // determine minimum and maximum k
                    if(k.size() == 1){

                        // we will change temp_k_max iff the number of points
                        // are larger than 10
                        if( temp_RData.size() > 10 ){
                            if ((int) (temp_RData.size() / 5) < k[0]){
                                temp_k_max = (int) (temp_RData.size() / 5);
                            } else {
                                temp_k_max = k[0];
                            }
                        }

                    } else {

                        // we will change temp_k_min and temp_k_max iff the number of points
                        // are larger than 10
                        if(temp_RData.size() > 10){
                            if ((int) (temp_RData.size() / 5) < k[0]){
                                temp_k_min = (int) (temp_RData.size() / 5);
                            } else {
                                temp_k_min = k[0];
                            }

                            if ((int) (temp_RData.size() / 5) < k[1]){
                                temp_k_max = (int) (temp_RData.size() / 5);
                            } else {
                                temp_k_max = k[1];
                            }

                        }
                    }

                    best_k = temp_k_min;
                    best_sil_score = (double)(-INT_MAX);

                    // find the optimal k using Silhouette score
                    for(int m=temp_k_min; m<=temp_k_max; m++){

                        // get cluster assignments
                        temp_classn = GetClusterAssignments(temp_pair_data, m);

                        // get Silhouette score for each observation
                        sil_scores = silhouette_score(temp_classn, dist_mat);

                        // obtain mean Silhouette score
                        mean_sil_score = GetSilMeanScore(sil_scores);

                        // obtain best sil score / best k
                        if(best_sil_score <= mean_sil_score){
                            best_sil_score = mean_sil_score;
                            best_k = m;
                        }
                    }

                    // finally apply GOC using the best k
                    // get cluster assignments
                    temp_classn = GetClusterAssignments(temp_pair_data, best_k);

                    // get discretized data
                    temp_Ddata = ApplyGOC(temp_classn, temp_RData, best_k);

                } else if(k.size() == nparents * children.size()){
                    // if size of k is nparents * nchildren, then it specifies the exact k for each combination

                    // get cluster assignments
                    temp_classn = GetClusterAssignments(temp_pair_data, k[effind]);

                    // get discretized data
                    temp_Ddata = ApplyGOC(temp_classn, temp_RData, k[effind]);
                }

                // make a contingency table
                temp_tab = tableCpp(temp_Ddata[0], temp_Ddata[1], -1, -1);

                // get stats
                stat = chisq(temp_tab, p_value, estimate);

                // store stats
                fill_stat_collector(sc[effind], i, c_indices[c], p_names[i], c_names[c_indices[c]], p_value, estimate);

            }

        } // parents

    } // children

}


/*
 * Master -- Parallel discrete coexpression network builder that works with continuous values.
 * 
 * Parameters
 * 
 * children: A continuous sample x gene matrix representing 
 * discrete levels of gene expression at given samples. This matrix
 * contains all child genes.
 * 
 * parents: A continuous sample x gene matrix representing 
 * discrete levels of gene expression at given samples. This matrix
 * contains all parent genes.
 * 
 * c_names: a vector of strings containing gene names for the children matrix.
 * 
 * p_names: a vector of strings containing gene names for the parents matrix.
 * 
 * k: a vector of integers that can contain the following:
 * i) k with size two: the minimum and maximum number of discrete levels allowed.
 * ii) k with size one: the maximum number of discrete levels allowed (minimum would be assumed to be 2).
 * iii) k with size n (where n is nparents x nchildren): the number of discrete levels to be used for each
 * interaction.
 * 
 * nthreads: an integer representing the numbers of thread to be used.
 * 
 */

vec<stat_collector> coexpnet(const frame<double> & children,
                             const frame<double> & parents,
                             const vec<std::string> & c_names,
                             const vec<std::string> & p_names,
                             const vec<int> & k,
                             int nthreads=-1) {

    // fix nthreads
    int true_threads;
    true_threads = (int)boost::thread::hardware_concurrency() - 1;

    if(nthreads > true_threads) {
        nthreads = true_threads;
    }

    if(nthreads < 2) {
        std::cout <<"WARNING: No multithreading enabled!"<<std::endl;
    }

    // create threads
    boost::thread_group threads;

    // dims for input
    int nchildren = children.size();
    int nparents = parents.size();

    // a vector of stat_collector data structure
    vec<stat_collector> sc(nparents*nchildren);

    // load balancing on number of children for multi-threads

    frame<int> c_indices;
    vec<int> temp_indices;
    int eq_child = (int)ceil((double)nchildren/nthreads); // total children for each thread

    for(int child_index=0; child_index!=nchildren; child_index++){
        if((child_index+1) % eq_child != 0){ // prepare child indices for each thread
            temp_indices.push_back(child_index);
        } else {
            temp_indices.push_back(child_index);
            c_indices.push_back(temp_indices); // move on to next thread
            temp_indices.clear();
        }
    }
    if(temp_indices.size()!=0){
        c_indices.push_back(temp_indices); // residual child indices
    }

    // Create threads for each set of children
    for(int i=0; i<c_indices.size(); i++){

        if(nthreads >= 2){ // multiple threads
            threads.create_thread(boost::bind(&coexpnet_thread,
                                              boost::cref(children),
                                              boost::cref(parents),
                                              boost::cref(c_names),
                                              boost::cref(c_indices[i]),
                                              boost::cref(p_names),
                                              boost::cref(k),
                                              boost::ref(sc)));
        } else { // single thread
            coexpnet_thread(children, parents, c_names, c_indices[i], p_names, k, sc);
        }

    }

    // join threads
    if(nthreads >= 2){
        threads.join_all();
    }

    return sc;
}

/*
 * Worker -- Parallel discrete coexpression network builder
 * Parameters
 * 
 * children: A discrete sample x gene matrix representing 
 * discrete levels of gene expression at given samples. This matrix
 * contains all child genes.
 * 
 * parents: A discrete sample x gene matrix representing 
 * discrete levels of gene expression at given samples. This matrix
 * contains all parent genes.
 * 
 * c_names: a vector of strings containing gene names for the children matrix
 * 
 * c_indices: a vector of integers representing the memory location in 'sc' that the 
 * worker process should work on.
 * 
 * p_names: a vector of strings containing gene names for the parents matrix
 * 
 * sc: An instance of stat_collector to gather results.
 * 
 */

void discrete_coexpnet_thread(const frame<int> & children,
                              const frame<int> & parents,
                              const vec<std::string> & c_names,
                              vec<int> c_indices,
                              const vec<std::string> & p_names,
                              vec<stat_collector> & sc){

    // dims for input parents
    int nparents = parents.size();
    int nchildren_t = c_indices.size();

    // temporary table
    frame<int> temp_tab;

    // temporary variables for stat collection
    int effind;
    double stat, estimate;
    ldouble p_value;

    for(int c=0; c<nchildren_t; c++) {

        // collect result for all parents except which c_names[c_indices[c]] == p_names[i]
        for(int i=0; i<nparents; i++){

            // effective index
            effind = (nparents * c_indices[c]) + i;

            if(c_names[c_indices[c]] == p_names[i]){

                // Not applicable -- store worst result
                fill_stat_collector(sc[effind], i, c_indices[c], p_names[i], c_names[c_indices[c]], 1, 0);

            } else {

                // make a contingency table
                temp_tab = tableCpp(parents[i], children[c_indices[c]], -1, -1);

                // get stats
                stat = chisq(temp_tab, p_value, estimate);

                // store stats
                fill_stat_collector(sc[effind], i, c_indices[c], p_names[i], c_names[c_indices[c]], p_value, estimate);

            }

        } // parents

    } // children

}


/*
 * Master -- Parallel discrete coexpression network builder
 * Parameters
 * 
 * children: A discrete sample x gene matrix representing 
 * discrete levels of gene expression at given samples. This matrix
 * contains all child genes.
 * 
 * parents: A discrete sample x gene matrix representing 
 * discrete levels of gene expression at given samples. This matrix
 * contains all parent genes.
 * 
 * c_names: a vector of strings containing gene names for the children matrix
 * 
 * p_names: a vector of strings containing gene names for the parents matrix
 * 
 * nthreads: an integer representing the numbers of thread to be used.
 * 
 */

vec<stat_collector> discrete_coexpnet(const frame<int> & children,
                                      const frame<int> & parents,
                                      const vec<std::string> & c_names,
                                      const vec<std::string> & p_names,
                                      int nthreads=-1) {

    // fix nthreads
    int true_threads = (int)boost::thread::hardware_concurrency()-1;

    if(nthreads > true_threads) {
        nthreads = true_threads;
    }

    if(nthreads < 2) {
        std::cout <<"WARNING: No multithreading enabled!"<<std::endl;
    }

    // create threads
    boost::thread_group threads;

    // dims for input
    int nchildren = children.size();
    int nparents = parents.size();

    // a vector of stat_collector data structure
    vec<stat_collector> sc(nparents*nchildren);

    // load balancing on number of children for multi-threads

    frame<int> c_indices;
    vec<int> temp_indices;
    int eq_child = (int)ceil((double)nchildren/nthreads); // total children for each thread

    for(int child_index=0; child_index!=nchildren; child_index++){
        if((child_index+1) % eq_child != 0){ // prepare child indices for each thread
            temp_indices.push_back(child_index);
        } else {
            temp_indices.push_back(child_index);
            c_indices.push_back(temp_indices); // move on to next thread
            temp_indices.clear();
        }
    }
    if(temp_indices.size()!=0){
        c_indices.push_back(temp_indices); // residual child indices
    }

    // Create threads for each set of children
    for(int i=0; i<c_indices.size(); i++){

        if(nthreads >= 2){ // multiple threads
            threads.create_thread(boost::bind(&discrete_coexpnet_thread,
                                              boost::cref(children),
                                              boost::cref(parents),
                                              boost::cref(c_names),
                                              boost::cref(c_indices[i]),
                                              boost::cref(p_names),
                                              boost::ref(sc)));
        } else { // single thread
            discrete_coexpnet_thread(children, parents, c_names, c_indices[i], p_names, sc);
        }

    }

    // join threads
    if(nthreads >= 2){
        threads.join_all();
    }

    return sc;
}

File: R/Coexpnet.R
# Coexpnet.R
#
# Created by Sajal Kumar
# Copyright (c) NMSU Song lab

#' A parallel co-expression network builder
#'
#' Builds a co-expression network based on the Pearson's chi-square test. 
#' The builder computes a chi-squared p-value and a cramer's V effect size for each pair of gene.
#' The builder provides options to discretize the expression profiles by either Ckmeans.1d.dp,
#' that is an optimal univariate clustering algorithm, or GridOnCluster, that discretizes genes
#' involved in a co-expression pattern based on their joint distribution. 
#'
#' @import Rcpp
#' @importFrom stats p.adjust mad sd
#' @useDynLib DiffXCoExpNet
#'
#' @param parent_expr a sample (rows) x gene (columns) continuous expression profile matrix containing 
#' the the parent genes. This matrix also represents one half of all desirable co-expression
#' gene pairs.
#' 
#' @param child_expr a sample (rows) x gene (columns) continuous expression profile matrix containing 
#' the the child genes. This matrix also represents the other half of all desirable co-expression
#' gene pairs. The parent_expr and child_expr can be the same expression matrix if
#' a complete graph between all genes in a dataset is required.
#' 
#' @param c_names a vector containing names of the child genes.
#' 
#' @param p_names a vector containing names of the parent genes. The p_names and c_names
#' vector can be the same if a parent_expr and child_expr are equivalent.
#'
#' @param method a string specifying which discretization method to use. The options are:
#' 'univariate' that will choose \code{Ckmeans.1d.dp} or 'joint' that will choose
#' \code{GridOnCluster}.
#'
#' @param k a range of clusters to be found on \code{parent_expr} and \code{child_expr}. 
#' The default minimum is 2 and the maximum is automatically computed to require atleast 
#' five points per cluster. If the \code{method} picked is \code{'joint'}, then a \code{Silhouette} 
#' score determines the number of \code{k}; if the \code{method} picked is \code{'univariate'}, 
#' then a \code{BIC} score determines the number of \code{k}. 
#' 
#' @param nthreads an integer specifying the number of parallel threads to be used. If not specified,
#' a non-parallel version of the code would be executed.
#'
#'
#' @return
#'
#' A data frame containing information on all possible parent x child co-expression pairs. 
#' The following items are reported about each gene pair:
#' \item{PID}{the one(1)-based column index in \code{parent_expr} matrix of the parent gene.}
#' 
#' \item{CID}{the one(1)-based column index in \code{child_expr} matrix of the child gene.}
#'
#' \item{PNAMES}{the name of the parent gene.}
#' 
#' \item{CNAMES}{the name of the child gene.}
#' 
#' \item{PVALUE}{the adjusted chi-squared p-value, depicting the statistical dependency between the gene pair.}
#' 
#' \item{ESTIMATE}{the cramer's v effect size, depicting the strength of the relationship between the 
#' gene pair.}
#' 
#' @author 
#' 
#' Ruby Sharma, Sajal Kumar and Mingzhou Song
#' 
#' @seealso
#'
#' See \pkg{Ckmeans.1d.dp} for discretizing univariate continuous data.
#' See \pkg{GridOnClusters} for joint discretization of continuous data.
#'
#' @export

Build_Coexpnet = function(parent_expr, 
                          child_expr, 
                          c_names, 
                          p_names,
                          method = c('univariate', 'joint'), 
                          k=NULL,
                          nthreads=NULL){
  
  # check if length of p_names is equal to the number of genes in parent_expr
  if(length(p_names) != ncol(parent_expr)){
    stop("The length of p_names should be equal to the number of genes in parent_expr.")
  }
  
  # check if length of c_names is equal to the number of genes in child_expr
  if(length(c_names) != ncol(child_expr)){
    stop("The length of c_names should be equal to the number of genes in child_expr.")
  }
  
  method = match.arg(method)
  
  # check if parent_expr and child_expr are equivalent
  if(identical(parent_expr, child_expr)){
    pc_expr_ident = TRUE
  } else {
    pc_expr_ident = FALSE
  }
  
  if(method == 'univariate'){
    
    # Discretize parent_expr using Ckmeans.1d.dp
    dis_parent_expr = DiscretizeData(parent_expr, nthreads)
    
    # if parent_expr and child_expr are equivalent then no need to discretize again
    if(pc_expr_ident){
      
      dis_child_expr = dis_parent_expr
    
    } else {
      
      # Discretize child_expr using Ckmeans.1d.dp
      dis_child_expr = DiscretizeData(child_expr, nthreads)
      
    }
    
    # find if there are any genes with only 1 level (they will removed)
    plevel = apply(dis_parent_expr, 2, max)
    remp = plevel == 1
    if (any(remp)){
      
      parent_expr = parent_expr[ , !remp]
      p_names = p_names[!remp]
      dis_parent_expr = dis_parent_expr[ , !remp]
      
      # child_expr will be treated similarly if identical parent_expr
      if(pc_expr_ident){
        
        child_expr = child_expr[ , !remp]
        c_names = c_names[!remp]
        dis_child_expr = dis_child_expr[ , !remp]
        
      }
    }
    
    # offset by 1
    dis_parent_expr = dis_parent_expr - 1
    dis_child_expr = dis_child_expr - 1
    
    if(pc_expr_ident){
      
      # build coexpression network with filtration
      coexpnet_result = Discrete_Coexpnet(children = dis_child_expr, 
                                          parents = dis_parent_expr,
                                          c_names = c_names,
                                          p_names = p_names,
                                          nthreads = nthreads,
                                          identical = TRUE)
      
    } else {
      
      # build coexpression network without filtration
      coexpnet_result = Discrete_Coexpnet(children = dis_child_expr, 
                                          parents = dis_parent_expr,
                                          c_names = c_names,
                                          p_names = p_names,
                                          nthreads = nthreads,
                                          identical = FALSE)
      
    }
    
  } else {
    
    # calculate maximum possible cluster
    max_possible = ceiling(sqrt((nrow(parent_expr)/5)))
    
    # check if user has provided k
    if(!is.null(k)){
     
      # check if the length of k is 2
      if(length(k) == 2){
        
        # check if the range is valid
        if(k[1] < 2 || k[2] < 2 || k[1] > max(max_possible,2) || k[2] > max(max_possible,2)){
          stop(paste0("Invalid range of supplied in k, the allowed range is between 2 to ", 
                      max(max_possible,2),"."))
        }
        
      } else {
        
        stop("Length of k should be 2. The first and second element should denote the minimum 
             and maximum number of clusters to consider, respectively.")
        
      }   
    } else { # determine the range of k
      
      k = c(2, max(max_possible, 2))
      if(k[2] == 2){
        k = 2
      }
      
    }
    
    if(pc_expr_ident){
      
      # build coexpression network with filtration
      coexpnet_result = Coexpnet(children = child_expr,
                                 parents = parent_expr,
                                 c_names = c_names,
                                 p_names = p_names,
                                 k = k,
                                 nthreads = nthreads,
                                 identical = TRUE)
      
    } else {
      
      # build coexpression network without filtration
      coexpnet_result = Coexpnet(children = child_expr,
                                 parents = parent_expr,
                                 c_names = c_names,
                                 p_names = p_names,
                                 k = k,
                                 nthreads = nthreads,
                                 identical = FALSE)
      
    }
    
  }
  
  # remove NA rows
  rem_entries = is.na(coexpnet_result$PID)
  
  if(any(rem_entries)){
    coexpnet_result = coexpnet_result[!rem_entries, ]
  }
  
  # adjust p-values
  coexpnet_result$PVALUE = p.adjust(coexpnet_result$PVALUE, method = "BH")
  
  # return coexpression network
  return(coexpnet_result)
}

